docs/testplan/Generic-hash-test-plan.md

```diff
@@ -0,0 +1,314 @@
+                "SRC_MAC",
+                "ETHERTYPE",
+                "IP_PROTOCOL",
+                "DST_IP",
+                "SRC_IP",
+                "L4_DST_PORT",
+                "L4_SRC_PORT",
+                "INNER_DST_MAC",
+                "INNER_SRC_MAC",
+                "INNER_ETHERTYPE",
+                "INNER_IP_PROTOCOL",
+                "INNER_DST_IP",
+                "INNER_SRC_IP",
+                "INNER_L4_DST_PORT",
+                "INNER_L4_SRC_PORT"
+            ]
+        }
+    }
+}
+```
+### 3.4 Supported topology
+The test supports t0 and t1 topologies, not supports dualtor topology.
+
+
+## 4. Test cases
+
+Notes: 
+  1. The hash field is randomly selected in the test cases. Currently these fields are tested: 'IN_PORT', 'SRC_MAC', 'DST_MAC', 'ETHERTYPE', 'VLAN_ID', 'IP_PROTOCOL', 'SRC_IP', 'DST_IP', 'L4_SRC_PORT', 'L4_DST_PORT', 'INNER_SRC_IP', 'INNER_DST_IP'.
+  2. DST_MAC, ETHERTYPE, VLAN_ID fields are only tested in lag hash test cases, because L2 traffic is needed to test these fields, and there is no ecmp hash when switching in L2.
+  3. IPv4 and IPv6 are covered in the test, but the versions(including the inner version when testing the inner fields) are randomly selected in the test cases.
+  4. For the inner fields, three types of encapsulations are covered: IPinIP, VxLAN and NVGRE. For the VxLAN packet, the default port 4789 and a custom port 13330 are covered in the test.
+
+### Test cases #1 - Verify the “show switch-hash capabilities” gets the supported hash fields.
+1. Get the supported hash fields via cli "show switch-hash capabilities"
+2. Check the fields are as expected`.
+
+### Test cases #2 - Verify when generic ecmp hash is configured, the traffic can be balanced accordingly.
+1. The test is using the default links and routes in a t0/t1 testbed.
+2. Randomly select a hash field and configure it to the ecmp hash list via cli "config switch-hash global ecmp-hash".
+3. Configure the lag hash list to exclude the selected field to verify the lag hash does not affect the hash result.
+4. Send traffic with changing values of the field under test from a downlink ptf port to uplink destination via multiple nexthops.
+5. Check the traffic is balanced over the nexthops.
+6. If the uplinks are portchannels with multiple members, check the traffic is not balanced over the members.
+
+### Test cases #3 - Verify when generic lag hash is configured, the traffic can be balanced accordingly.
+1. The test is using the default links and routes in a t0/t1 testbed, and only runs on setups which have multi-member portchannel uplinks.
+2. Randomly select a hash field and configure it to the lag hash list via cli "config switch-hash global lag-hash".
+3. Configure the ecmp hash list to exclude the selected field to verify the ecmp hash does not affect the hash result.
+4. If the hash field is DST_MAC, ETHERTYPE or VLAN_ID, change the topology to allow L2 switching and send L2 traffic in the next step.
```

Comment:
@congh-nvidia what does it mean "change topology"? IMHO, it doesn't efficient during test execution...

Reply:
The community testbeds only have L3 configurations, if we want to test L2 traffic, need to do configure some L2 ports and vlan, it doesn't cost much time. No need to re-deploy sonic topology.

Reply:
@congh-nvidia is it reasonable to have a separate test case for that? Just to not mix a lot of stuff for the same scenario.

Reply:
To save the execution time, we are not aiming to cover all the possible scenarios and cover all fields in a single run, the hash field is generated randomly. That's why all the fields are tested in a single test case.